R/read_redcap.R

```diff
@@ -72,20 +73,83 @@ read_redcap <- function(redcap_uri,
                         forms = NULL,
                         export_survey_fields = TRUE,
                         suppress_redcapr_messages = TRUE) {
-
   check_arg_is_character(redcap_uri, len = 1, any.missing = FALSE)
+  check_arg_is_character(token, len = 1, any.missing = FALSE)
   check_arg_is_valid_token(token)
   check_arg_choices(raw_or_label, choices = c("label", "raw"))
   check_arg_is_character(forms, min.len = 1, null.ok = TRUE, any.missing = FALSE)
   check_arg_is_logical(export_survey_fields, len = 1, any.missing = FALSE)
   check_arg_is_logical(suppress_redcapr_messages, len = 1, any.missing = FALSE)
 
   # Load REDCap Metadata ----
-  db_metadata <- redcap_metadata_read(
-    redcap_uri = redcap_uri,
-    token = token,
-    verbose = !suppress_redcapr_messages
-  )$data %>%
+  # Capture unexpected metadata API call errors
+  try_fetch(
+    {
+      call <- current_call()
+
+      out <- redcap_metadata_read(
+        redcap_uri = redcap_uri,
+        token = token,
+        verbose = !suppress_redcapr_messages
+      )
+
+      # Throw an error if the API calling function finished successfully
+      # but the API call itself was not successful
+      if (out$success == FALSE) {
+        abort(
+          class = c("redcapr_api_call_success_false", "REDCapTidieR_cond")
+        )
+      }
+    },
+    error = function(cnd) {
+      error_message <- c("x" = "The REDCapR metadata export operation was not successful.")
+
+      # Show original error?
+      parent <- NULL
+
+      error_class <- c("metadata_api_call_failed", "REDCapTidieR_cond")
+
+      if (exists("out") && out$status_code == 403) {
```

Comment:
This seems slightly fragile since we're looking outside of the current function env for an object called `out`. An alternative would be to:

1. Pass `out$status_code` as error metadata on L98-102 like

```r
abort(
    class = c("redcapr_api_call_success_false", "REDCapTidieR_cond"),
    status_code = out$status_code
)
```

Then you could safely reference it as `cnd$status_code`

2. Move L112-125 to a `try_fetch` handler for `redcapr_api_call_success_false` above the one for `error` so the code would look like:

```r
try_fetch({
    ...
    },
    redcapr_api_call_success_false = function(cnd) {
        # L112-125 logic
        if (cnd$status_code == 403) {
        ...
        } else if ...
    },
    error = function(cnd) {
        # L125-139 logic
    }
)

Reply:
How awesome, I had no idea that you can pass arbitrary metadata inside the `cnd` object! Implemented mostly as suggested, except I kept all the error handling in one function instead of breaking into two, for now, for simplicity's sake, but it's a great point that we can handle each error class with a separate handler function. Anyhow, I think we can later decide how to refactor.